lighthouse-core/test/audits/performance-budget-test.js

```diff
@@ -0,0 +1,164 @@
+            resourceType: 'image',
+            budget: 1000,
+          },
+        ],
+        resourceCounts: [
+          {
+            resourceType: 'script',
+            budget: 0,
+          },
+          {
+            resourceType: 'image',
+            budget: 1000,
+          },
+        ],
+      }];
+    });
+
+    it('includes table columns for requet & file size overages', async () => {
+      const result = await ResourceBudgetAudit.audit(artifacts, context);
+      expect(result.details.headings).toHaveLength(4);
+    });
+
+    it('table item information is correct', async () => {
+      const result = await ResourceBudgetAudit.audit(artifacts, context);
+      const item = result.details.items[0];
+      expect(item.label).toBeDisplayString('Script');
+      expect(item.requestCount).toBe(2);
+      expect(item.size).toBe(60);
+      expect(item.sizeOverBudget).toBe(60);
+      expect(item.countOverBudget).toBeDisplayString('2 requests');
+    });
+
+    describe('request & transfer size overage', () => {
+      it('are displayed', async () => {
+        const result = await ResourceBudgetAudit.audit(artifacts, context);
+        const scriptRow = result.details.items.find(r => r.resourceType === 'script');
+        expect(scriptRow.sizeOverBudget).toBe(60);
+        expect(scriptRow.countOverBudget).toBeDisplayString('2 requests');
+      });
+
+      it('are empty for passing budgets', async () => {
+        const result = await ResourceBudgetAudit.audit(artifacts, context);
+        const imageRow = result.details.items.find(r => r.resourceType === 'image');
+        expect(imageRow.sizeOverBudget).toBeUndefined();
+        expect(imageRow.countOverBudget).toBeUndefined();
+      });
+
+      it('convert budgets from kilobytes to bytes during calculations', async () => {
+        context.settings.budgets = [{
+          resourceSizes: [
+            {
+              resourceType: 'document',
+              budget: 20,
+            },
+          ],
+        }];
+        const result = await ResourceBudgetAudit.audit(artifacts, context);
+        expect(result.details.items[0].siveOverBudget).toBeUndefined();
+      });
+    });
+
+    it('only includes rows for resource types with budgets', async () => {
+      const result = await ResourceBudgetAudit.audit(artifacts, context);
+      expect(result.details.items).toHaveLength(2);
+    });
+
+    it('sorts rows by descending file size overage', async () => {
+      context.settings.budgets = [{
+        resourceSizes: [
+          {
+            resourceType: 'document',
+            budget: 0,
+          },
+          {
+            resourceType: 'script',
+            budget: 0,
+          },
+          {
+            resourceType: 'image',
+            budget: 0,
+          },
+        ],
+      }];
+      const result = await ResourceBudgetAudit.audit(artifacts, context);
+      const items = result.details.items;
+      items.slice(0, -2).forEach((item, index) => {
```

Comment:
we're not putting `third-party` last in this audit, so shouldn't use the `slice()`?

Reply:
Changed... but to `slice(0,-1)`. `slice()` or something else is needed so that the last item isn't compared to the next item, which doesn't exist.

Reply:
oh right, good point